File: src/player.py
from enum import Enum, unique
from card import ActionCard, Deck

class Player:
    """
    Is part of the @class State object. It remembers the state of a player (resources, etc.) at a certain point in time.
    """
    startingAbilityDeck = [ ActionCard.DEFENSE ] * 2 +\
        [ ActionCard.HARDATK ] * 4 +\
        [ ActionCard.SOFTATK ] * 4

    def __init__(self, abilityDeck = Deck.create([])):
        self.resources = []
        self.actions = []
        self.tokens = []
        self.actionDeck = Deck.create(Player.startingAbilityDeck)
        self.abilityDeck = abilityDeck

    def serialize(self):
        return {
            "actionDeckSize" : self.actionDeck.sizeconfig()
        }

    def useActionCard(self, cardValue):
        self.actions.remove(cardValue)
        self.actionDeck.discard(cardValue)

    def drawAction(self):
        cardDrawn = self.actionDeck.draw()
        self.actions.append(cardDrawn)
        #print('Drew', cardDrawn, ', now have', self.actions)
        return cardDrawn

    def drawResource(self, resourceDeck):
        cardDrawn = resourceDeck.draw()
        self.resources.append(cardDrawn)
        return cardDrawn

    def discard(self, card):
        if card not in self.actions:
            print('Self.actions:', self.actions)
            print('Card', card)
        self.actions.remove(card)
        self.actionDeck.discard(card)

    #virtual
    def getAbilityDeck(self):
        return self.abilityDeck

File: src/player.cpp
#include "player.hpp"
#include "deck.hpp"
#include "constants.hpp"
#include "random.hpp"
#include <ostream>
#include <algorithm>
#include <cassert>

Player::Player(Generator generator) {
    deck = Deck<CombatDeckCard>::create(startingAbilityDeck, generator);
}

std::variant<ActionCard, Faction> Player::drawCard() {
    std::variant<ActionCard, Faction> cardDrawn = this->deck.draw();
    if(cardDrawn.index() == 0) {
        actions.insert(actions.end(), std::get<0>(cardDrawn));
    } else {
        resources.insert(resources.end(), std::get<1>(cardDrawn));
    }
    return cardDrawn;
}

std::ostream& operator<<(std::ostream& os, const CombatDeckCard& card){
    if(card.index() == 0) return os << std::get<0>(card);
    else return os << std::get<1>(card);
}

void Player::discard(CombatDeckCard card) {
    if(card.index() == 0) discardAction(std::get<ActionCard>(card));
    else discardResource(std::get<Faction>(card));
}

void Player::discardAction(unsigned int iCard) {
    assert(iCard < actions.size());
    auto card = actions[iCard];
    actions.erase(actions.begin() + iCard);
    deck.discard(card);
}

void Player::discardAction(ActionCard card) {
    auto cardPosition = std::find(actions.begin(), actions.end(), card);
    assert(cardPosition != actions.end());
    actions.erase(cardPosition);
    deck.discard(card);
}

void Player::discardResource(unsigned int iCard) {
    assert(iCard < resources.size());
    auto card = resources[iCard];
    actions.erase(actions.begin() + iCard);
    deck.discard(card);
}

void Player::discardResource(Faction card) {
    auto cardPosition = std::find(resources.begin(), resources.end(), card);
    assert(cardPosition != resources.end());
    resources.erase(cardPosition);
    deck.discard(card);
}
